san/impl/derive_type.cpp

```diff
@@ -0,0 +1,256 @@
+
+void DeriveType::visit(ArrayType* node) {
+
+    if (node->length) {
+        node->length->accept(*this);
+        // any primitive type can be convered to the int
+        if (!typeIsBase(node->length->type)) {
+            error(node->begin, "invalid array length type, should be integer");
+        }
+    } else if (!m_inRoutineParams) {
+        error(node->begin, "length of array should be always defined");
+    } // length can be absent in case of routine parameter
+    node->elementType->accept(*this);
+    if (m_searchArray) {
+        m_arrayInnerType = node->elementType;
+    }
+}
+
+void DeriveType::visit(RecordType* node) {
+    for (auto field : node->fields) {
+        field->accept(*this);
+        if (m_searchRecord && field->name == m_recordField) {
+            m_recordInnerType = field->type;
+        }
+    }
+}
+
+void DeriveType::visit(VariableDecl* node) {
+
+    // check type
+    if (node->type != nullptr) {
+        node->type->accept(*this);
+    }
+    // check/evaluate expression value
+    if (node->initialValue != nullptr) {
+        node->initialValue->accept(*this);
+    }
+    // assign initial value
+    if (node->type == nullptr) {
+        node->type = node->initialValue->type;
+    }
+}
+
+void DeriveType::visit(TypeDecl* node) { node->type->accept(*this); }
+
+void DeriveType::visit(Body* node) {
+
+    // to check array length type
+    for (auto type : node->types) {
+        type->type->accept(*this);
+    }
+    // check variables with both type and initial value
+    for (auto var : node->variables) {
+        var->accept(*this);
+    }
+    // routine has an expression
+    for (auto statement : node->statements) {
+        statement->accept(*this);
+    }
+}
+
+void DeriveType::visit(Statement*) {}
+
+void DeriveType::visit(ReturnStatement*) {}
+
+void DeriveType::visit(Assignment* node) { node->rhs->accept(*this); }
+
+void DeriveType::visit(WhileLoop* node) {
+    node->condition->accept(*this);
+    node->body->accept(*this);
+}
+
+void DeriveType::visit(ForLoop* node) {
+    // check that loop var is of type int
+    node->rangeFrom->accept(*this);
+    if (!typeIsBase(node->rangeFrom->type)) {
+        error(node->rangeFrom->begin, "invalid type of the beginning of range");
+    }
+
+    node->rangeTo->accept(*this);
+    if (!typeIsBase(node->rangeTo->type)) {
+        error(node->rangeTo->begin, "invalid type of the beginning of range");
+    }
+
+    node->loopVar->type = std::make_shared<ast::IntegerType>();
+    node->body->accept(*this);
+}
+
+void DeriveType::visit(IfStatement* node) {
+    node->condition->accept(*this);
+    // check condition is of type boolean
+    node->ifBody->accept(*this);
+    if (node->elseBody != nullptr) {
+        node->elseBody->accept(*this);
+    }
+}
+
+void DeriveType::visit(Expression*) {}
+
+void DeriveType::visit(UnaryExpression* node) {
+    node->operand->accept(*this);
+    node->type = node->operand->type;
```

Comment:
Not necessarily; it depends on the operator. In the case of `not`, for example, the resulting type is `boolean`